src/vm/analysis/definition_sorter/mod.rs

```diff
@@ -94,11 +103,9 @@ impl <'a> DefinitionSorter {
                                 DefineFunctions::PrivateFunction | DefineFunctions::Constant |
                                 DefineFunctions::PublicFunction | DefineFunctions::PersistedVariable |
                                 DefineFunctions::ReadOnlyFunction => {
-                                    // Args: [(define-name-and-types), ...]: ignore 1st arg
-                                    if function_args.len() > 1 {
-                                        for expr in function_args[1..function_args.len()].into_iter() {
-                                            self.probe_for_dependencies(expr, tle_index)?;
-                                        }
+                                    // Args: [(define-name-and-types), ...]
+                                    for expr in function_args.into_iter() {
```

Comment:
Because generics are only permitted in function definitions, we should separate this branch into `Constant | PersistedVariable`, which can be handled as before, and a branch for `PrivateFunction | PublicFunction | ReadOnlyFunction`, which needs to handle generic parameters.

For handling the parameters, I think they may need to be handled with a special case rather than just calling `probe_for_dependencies` -- the current error from this:

```
        (define-private (function-1 (function-2 int))
            (+ 1 2))
        (define-private (function-2 (function-1 int))
            (+ 1 2))
```

is a circular reference, and throws a circular reference error, but that's kind of a strange error, because the real error should be something like "NameAlreadyUsed" (since we don't allow shadowing of function names). So maybe a function like `probe_for_dependencies_in_define_args` should get called on arg1.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/384af248c1b84c523f8ba69f7c8d962165e7ef6c